alf/environments/wrappers.py

```diff
@@ -12,187 +12,166 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import collections
-from collections import OrderedDict, deque
-from copy import deepcopy
-
-import gin
+import copy
+from collections import deque
 import gym
 import numpy as np
 import cv2
-
-from absl import logging
-
+import gin
 from tf_agents.environments import wrappers
 from tf_agents.trajectories.time_step import StepType
+from alf.utils import common
+
+
+def traverse_transform_space(space, levels, field, func):
+    """Transform space
+
+    Args:
+         space (gym.Space): space to be transformed
+         levels (list[str]): levels
+         field (str): filed name, a multi-step path denoted by "A.B.C"
+         func (Callable): transform func
+    """
+    if not levels:
+        return func(space, field)
+
+    assert isinstance(space, gym.spaces.Dict)
+    level = levels[0]
+
+    new_val = copy.deepcopy(space)
+    new_val.spaces[level] = traverse_transform_space(
+        space=space.spaces[level], levels=levels[1:], field=field, func=func)
+    return new_val
 
 
 @gin.configurable
-class FrameStack(gym.Wrapper):
+class BaseObservationWrapper(gym.ObservationWrapper):
+    """Base observation Wrapper
+
+    BaseObservationWrapper provide basic functions and generic interface for transformation.
+
+    The key interface functions are:
+    1. transform_space(): transform space.
+    2. transform_observation(): transform observation.
+    """
+
+    def __init__(self, env, fields=None):
+        """
+        Args:
+            env (gym.Env): the gym environment
+            fields (list[str]): fields to be applied transformation
+        """
+        super().__init__(env)
+        paths = []
+        for field in fields or [""]:
+            # remove '' in the path
+            paths.append(([step for step in field.split(".") if step], field))
+        self._paths = paths
+        observation_space = env.observation_space
+        for levels, field in self._paths:
+            observation_space = traverse_transform_space(
+                space=observation_space,
+                levels=levels,
+                field=field,
+                func=self.transform_space)
+        self.observation_space = observation_space
+
+    def observation(self, observation):
+        for levels, field in self._paths:
+            observation = common.traverse_transform_observation(
+                obs=observation,
+                levels=levels,
+                field=field,
+                func=self.transform_observation)
+        return observation
+
+    def transform_space(self, observation_space, field=None):
```

Comment:
`field` seems not necessary if this function is always called on a leaf

Reply:
field is necessary in some case , from `FrameStack` , field represents a unique key for a queue (many queues ,each for a field ) that handle the stacking of this field

Reply:
Oh I see